src/early-types.lisp

```diff
@@ -53,17 +53,67 @@
 (defvar *type-parsers* (make-hash-table :test 'equal)
   "Hash table of defined type parsers.")
 
+(define-condition cffi-error (error)
+  ())
+
+(define-condition type-related-error (cffi-error)
+  ((type-name :initarg :type-name
+              :initform (error "Must specify TYPE-NAME.")
+              :accessor type-related-error/type-name)
+   (namespace :initarg :namespace
+              :initform :default
+              :accessor type-related-error/namespace)))
+
+(defun type-related-error/compund-name (e)
+  (let ((name (type-related-error/type-name e))
+        (namespace (type-related-error/namespace e)))
+    (if (eq namespace :default)
+        name
+        `(,namespace ,name))))
+
+(define-condition simple-type-related-error (simple-error type-related-error)
+  ())
+
+(defun simple-type-related-error (type-name namespace format-control &rest format-arguments)
+  (error 'simple-type-related-error
+         :type-name type-name :namespace namespace
+         :format-control format-control :format-arguments format-arguments))
+
+(define-condition undefined-type-error (type-related-error)
+  ()
+  (:report (lambda (e stream)
+             (format stream "Unknown CFFI type ~S" (type-related-error/compund-name e)))))
+
+(defun undefined-type-error (type-name &optional (namespace :default))
+  (error 'undefined-type-error :type-name type-name :namespace namespace))
+
+;; TODO this is not according to the C namespace rules,
+;; see bug: https://bugs.launchpad.net/cffi/+bug/1527947
+(deftype c-namespace-name ()
+  '(member :default :struct :union))
+
+;; for C namespaces read: https://stackoverflow.com/questions/12579142/type-namespace-in-c
+;; (section 6.2.3 Name spaces of identifiers)
+;; NOTE: :struct is probably an unfortunate name for the tagged (?) namespace
 (defun find-type-parser (symbol &optional (namespace :default))
-  "Return the type parser for SYMBOL."
+  "Return the type parser for SYMBOL. NAMESPACE is either :DEFAULT (for
+variables, functions, and typedefs) or :STRUCT (for structs, unions, and enums)."
+  (check-type symbol (and symbol (not null)))
+  (check-type namespace c-namespace-name)
   (or (gethash (cons namespace symbol) *type-parsers*)
-      (if (eq namespace :default)
-          (error "unknown CFFI type: ~S." symbol)
-          (error "unknown CFFI type: (~S ~S)." namespace symbol))))
+      (undefined-type-error symbol namespace)))
 
 (defun (setf find-type-parser) (func symbol &optional (namespace :default))
   "Set the type parser for SYMBOL."
+  (check-type symbol (and symbol (not null)))
+  (check-type namespace c-namespace-name)
+  ;; TODO Shall we signal a redefinition warning here?
```

Comment:
i guess by checking if the hashtable value will be overwritten or assigned for the first time.

note: i may be unaware of some compile-time/load-time/execute headaches, though.